### PR TITLE
ts, ui, sql, ccl: add metrics, timeseries graph for feature flags

### PR DESCRIPTION
**Pull request by angelapwen**
This change adds metrics for statement denials due to feature
flags set via cluster settings, and displays a timeseries graph
tracking this metric on the SQL Dashboard of DB Console.

Feature Flag Denials Timeseries Graph on DB Console

Release note (ui change): Adds a timeseries graph indicating
statement denials due to feature flags on the SQL Dashboard of DB
Console.

### CODE REVIEW

**Comment by angelapwen** [comment]
___
*[pkg/featureflag/feature_flags.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjZVH7xuFEwhHFLTW:-MNfjZVH7xuFEwhHFLTX:b-wgvrrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L43)):*
> ```Go
> 	if enabled := s.Get(sv); !enabled {
> 		telemetry.Inc(sqltelemetry.FeatureDeniedByFeatureFlagCounter)
> 		accessories.Metrics().FeatureDenialMetric.Inc(1)
> ```

Currently this line has a nil pointer panic, because the `Metrics()` returned is actually not instantiated. Looking for a good point to instantiate it — somewhere right when the server starts?

**Comment by angelapwen** [comment]
___
*[pkg/featureflag/feature_flags.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjp8T8RHwme3PY8Qb:-MNfjp8UB_M2NCYutrhQ:bi06e7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L91)):*
> ```Go
> // FeatureFlagAccessories is an interface used to pass required values from
> // ExecutorConfig to check if a feature flag is enabled.
> type FeatureFlagAccessories interface {
> ```

Not sure about this name :grimacing:

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @miretskiy, and @otan)

___
*[pkg/featureflag/feature_flags.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjZVH7xuFEwhHFLTW:-MNfqTC02ZNkOkX27C2e:b-5xr8hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L43)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Currently this line has a nil pointer panic, because the `Metrics()` returned is actually not instantiated. Looking for a good point to instantiate it — somewhere right when the server starts?
</blockquote></details>

Yep. https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server_sql.go#L450 is fine.
___
*[pkg/featureflag/feature_flags.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfqeIe-JUfHoKvMG5Q:-MNfqeIe-JUfHoKvMG5R:b-fgp956) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L74)):*
> ```Go
> // this struct.
> type FeatureFlagMetrics struct {
> 	FeatureDenialMetric *metric.Counter
> ```

this should be private
___
*[pkg/featureflag/feature_flags.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjp8T8RHwme3PY8Qb:-MNfqiVFCAD3pJR9XEvk:b-bgpyxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L91)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Not sure about this name :grimacing: 
</blockquote></details>

maybe `FeatureFlagConfig`

**Comment by otan** [comment]
___
*[pkg/featureflag/feature_flags.go, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjZVH7xuFEwhHFLTW:-MNfqxe70tBFxdo1bIsP:bxy6kzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L43)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yep. https://github.com/cockroachdb/cockroach/blob/master/pkg/server/server_sql.go#L450 is fine.
</blockquote></details>

Maybe also add a `== nil` check, and if it is `nil` log a warning (no V() check needed) so that this function has no chance of crashing, and return the operator error anyway.

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen and @miretskiy)

___
*[pkg/featureflag/feature_flags.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfrEy5CQHO5nBQMPkx:-MNfrEy5CQHO5nBQMPky:bt9mz5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L93)):*
> ```Go
> type FeatureFlagAccessories interface {
> 	SV() *settings.Values
> 	Metrics() *FeatureFlagMetrics
> ```

maybe FeatureFlagMetrics(), so that we're not making this super generic on the ExecCfg

**Review by angelapwen — COMMENTED** [review]
Addressed comments, still need to do testing to make sure everything shows up ok on the DB Console!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-:-MNiz4MY5Bq_Pz-iMEat:b8imrt8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @miretskiy, and @otan)

___
*[pkg/featureflag/feature_flags.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfqeIe-JUfHoKvMG5Q:-MNitg1GECfNjfQWePSY:bkb9rsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L74)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should be private
</blockquote></details>

Got it — I assume you mean the `featureDenialMetric` and not the struct itself.
___
*[pkg/featureflag/feature_flags.go, line 91 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfjp8T8RHwme3PY8Qb:-MNitT6E-OMGu_Ux3a1r:b2nhlkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L91)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

maybe `FeatureFlagConfig`
</blockquote></details>

Ok!
___
*[pkg/featureflag/feature_flags.go, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfrEy5CQHO5nBQMPkx:-MNitS32A-0zFokYjKxf:bxxx4g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L93)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

maybe FeatureFlagMetrics(), so that we're not making this super generic on the ExecCfg
</blockquote></details>

That makes sense!
___
*[pkg/featureflag/feature_flags.go, line 46 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNiynwHCOvU9bTUCKZ8:-MNiynwHCOvU9bTUCKZ9:bor4vzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8de1661a3422673dba14d27d5f808ba2db45092/pkg/featureflag/feature_flags.go#L46)):*
> ```Go
> 			log.Warningf(
> 				ctx,
> 				"executorConfig.FeatureFlagMetrics is uninitiated; feature %s was attempted but disabled via cluster settings",
> ```

Not sure what this warning text should be: should it have both why it failed and that the feature was attempted like this?
___
*[pkg/sql/exec_util.go, line 750 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNiywDe4gfP4ZHhnM8x:-MNiywDe4gfP4ZHhnM8y:b-slvfvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8de1661a3422673dba14d27d5f808ba2db45092/pkg/sql/exec_util.go#L750)):*
> ```Go
> 
> // GetFeatureFlagMetrics returns the value of the FeatureFlagMetrics struct.
> func (cfg *ExecutorConfig) GetFeatureFlagMetrics() *featureflag.FeatureFlagMetrics {
> ```

Couldn't make the method & variable have the same name so added `Get`: not sure if this aligns with Cockroach/Go-style guidelines

**Comment by angelapwen** [comment]
___
*[pkg/featureflag/feature_flags.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNfqeIe-JUfHoKvMG5Q:-MNje0E--8HEEWN_QG-4:b-t93106) ([raw file](https://github.com/cockroachdb/cockroach/blob/d160b17dacd2683ba6977f878484ffd615182888/pkg/featureflag/feature_flags.go#L74)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Got it — I assume you mean the `featureDenialMetric` and not the struct itself.
</blockquote></details>

Commenting for posterity — had to be public so that it could be exported to the metric registry.

**Review by dhartunian — COMMENTED** [review]
> Reviewed 44 of 48 files at r1, 1 of 4 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @dhartunian, @miretskiy, @otan, and @taroface)

___
*[pkg/sql/exec_util.go, line 750 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNiywDe4gfP4ZHhnM8x:-MNjWuy87E10PZAYYADI:b3vwaj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8de1661a3422673dba14d27d5f808ba2db45092/pkg/sql/exec_util.go#L750)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Couldn't make the method & variable have the same name so added `Get`: not sure if this aligns with Cockroach/Go-style guidelines
</blockquote></details>

Can you let folks use `cfg.FeatureFlagMetrics` directly without requiring a getter function at all?
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNjjZQp8OqC90OGdaHE:-MNjjZQp8OqC90OGdaHF:b-sdiwsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5977f4d87bc43d6bbd6d8019534c7aa0085574/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx#L299)):*
> ```XML
>       title="Feature Flag Denial"
>       sources={nodeSources}
>       tooltip={`The number of statements denied by a feature flag ${tooltipSelection}.`}
> ```

@taroface can you take a look at the graph name and tooltip here. My instinct is that we might need some more background info here.

**Comment by angelapwen** [comment]
___
*[pkg/sql/exec_util.go, line 750 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNiywDe4gfP4ZHhnM8x:-MNjo1wo4dK7c1z3YKFn:bypvafi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8de1661a3422673dba14d27d5f808ba2db45092/pkg/sql/exec_util.go#L750)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you let folks use `cfg.FeatureFlagMetrics` directly without requiring a getter function at all?
</blockquote></details>

Normally, yes, but I added an interface (now called `FeatureFlagConfig`) to pass just these two variables (settings values and the feature flag metrics) through which requires getter functions for each one (to the best of my knowledge!)

**Review by dhartunian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @dhartunian, @miretskiy, @otan, and @taroface)

___
*[pkg/sql/exec_util.go, line 750 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNiywDe4gfP4ZHhnM8x:-MNjrQJC3cbGOPPWe0gj:b-8xk1d0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8de1661a3422673dba14d27d5f808ba2db45092/pkg/sql/exec_util.go#L750)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Normally, yes, but I added an interface (now called `FeatureFlagConfig`) to pass just these two variables (settings values and the feature flag metrics) through which requires getter functions for each one (to the best of my knowledge!)
</blockquote></details>

ah, gotcha :thumbsup: thx

**Comment by angelapwen** [comment]
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNjjZQp8OqC90OGdaHE:-MNk5MctCXVvPrk2MP9L:b-qjaa4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5977f4d87bc43d6bbd6d8019534c7aa0085574/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx#L299)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

@taroface can you take a look at the graph name and tooltip here. My instinct is that we might need some more background info here.
</blockquote></details>

Thank you both! Let me know what other information you you need from me/what I might want to add here. 

Also, I changed it to "Feature Flag Denials" to be more in line with the plural titles like "Schema Changes" :smile:

**Review by otan — APPROVED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @miretskiy, @otan, and @taroface)

___
*[pkg/ts/catalog/catalog_generator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNzn9aW8SPvVlAMynkQ:-MNzn9aW8SPvVlAMynkR:b-dftkq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e4de873fc9e2559d89a7932039fbbb6a072325a/pkg/ts/catalog/catalog_generator.go#L377)):*
> ```Go
> 
> 	// LongTitles look like "SQL Layer | SQL | Connections".
> 	// CollectionTitless look like "sql-layer-sql-connections".
> ```

nit: Titles

File: pkg/featureflag/feature_flags.go [hunk 1]
```diff
@@ -30,10 +31,29 @@ const FeatureFlagEnabledDefault = true
...
+
 	if enabled := s.Get(sv); !enabled {
+		// Increment telemetry counter for feature flag denial.
 		telemetry.Inc(sqltelemetry.FeatureDeniedByFeatureFlagCounter)
+
+		// Increment metrics counter for feature flag denial.
+		if config.GetFeatureFlagMetrics() == nil {
+			log.Warningf(
+				ctx,
+				"executorConfig.FeatureFlagMetrics is uninitiated; feature %s was attempted but disabled via cluster settings",
```
Comment (ericharmeling) [inline pkg/featureflag/feature_flags.go:46]: Would there be a way to call out the associated cluster setting here (idk if that information is provided by `featureName` or elsewhere)?

Comment (angelapwen) [reply pkg/featureflag/feature_flags.go:46]: At the moment no, the `featureName` is not linked to the associated cluster setting. I can see how that would be useful. I know that @vy-ton is considering the best way to document which cluster settings may block which statements (as in the future there may be multiple categories that one statement can fall under).

**Review by taroface — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @dhartunian, @miretskiy, @otan, and @vy-ton)

___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNjjZQp8OqC90OGdaHE:-MO3fwuM-RPPeVdJGs0J:bytnuc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5977f4d87bc43d6bbd6d8019534c7aa0085574/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx#L299)):*
Hmm, is "feature flag" a technical term that refers to the value of a cluster setting? I've seen the phrase take a different meaning in other contexts and I don't think our docs use it together with cluster settings. I would be inclined to spell it out in the tooltip:

> The total number of statements denied per second [across all nodes / on node \<node\>] due to the value of a cluster setting.

This format would be consistent with the tooltips on the other graphs, where "across all nodes" vs. "on node \<node\>" (\<node\> being the node ID) depends on whether you're looking at the cluster or a specific node. 

Perhaps "cluster setting" could be linked to https://www.cockroachlabs.com/docs/stable/cluster-settings.html (there's a dynamic way to link to the docs version that corresponds to the Cockroach version being used, but I don't know how).

For the graph name, something like `Statement Denials: Cluster Settings` is clearer to me, and leaves room for graphs tracking other kinds of statement denials, but tell me if "Feature Flag Denials" is likely self-evident.

**Comment by angelapwen** [comment]
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNjjZQp8OqC90OGdaHE:-MOIslM5EfIP7kZoaown:b-c31up5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5977f4d87bc43d6bbd6d8019534c7aa0085574/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx#L299)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

Hmm, is "feature flag" a technical term that refers to the value of a cluster setting? I've seen the phrase take a different meaning in other contexts and I don't think our docs use it together with cluster settings. I would be inclined to spell it out in the tooltip:

> The total number of statements denied per second [across all nodes / on node \<node\>] due to the value of a cluster setting.

This format would be consistent with the tooltips on the other graphs, where "across all nodes" vs. "on node \<node\>" (\<node\> being the node ID) depends on whether you're looking at the cluster or a specific node. 

Perhaps "cluster setting" could be linked to https://www.cockroachlabs.com/docs/stable/cluster-settings.html (there's a dynamic way to link to the docs version that corresponds to the Cockroach version being used, but I don't know how).

For the graph name, something like `Statement Denials: Cluster Settings` is clearer to me, but tell me "Feature Flag Denials" is likely self-evident.
</blockquote></details>

Thank you for the suggestions @taroface! I added some detail to the tooltip: do you think the extra detail is clear enough now?

timeseries

I found the method in the database that allows for dynamic linking the docs version that corresponds to the right version but I'm not really sure how to integrate it into the `.tsx` file — @dhartunian does this require passing the value through a proto to the React components?

**Review by dhartunian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @dhartunian, @miretskiy, @otan, @taroface, and @vy-ton)

___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx, line 299 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57533#-MNjjZQp8OqC90OGdaHE:-MOJ0URg6y-NUlj5H4FR:b-l6ejbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5977f4d87bc43d6bbd6d8019534c7aa0085574/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx#L299)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Thank you for the suggestions @taroface! I added some detail to the tooltip: do you think the extra detail is clear enough now?

timeseries

I found the method in the database that allows for dynamic linking the docs version that corresponds to the right version but I'm not really sure how to integrate it into the `.tsx` file — @dhartunian does this require passing the value through a proto to the React components?
</blockquote></details>

@angelapwen what Ryan is referring to, I believe, are the helper functions in `pkg/ui/src/util/docs.ts` which let you link to pages in our docs in a reliable way. The `docsURL` function in there is already configured to use the DB version in the URL prefix automatically. Let me know if you have further usage questions.

**Comment by angelapwen** [comment]
Updated with @taroface 's great comments on making the graph more comprehensible! See PR body for updated screenshot of the graph and tooltip. Thanks @dhartunian for teaching me a bit more about React :sparkling_heart:

**Comment by angelapwen** [comment]
bors r=otan

**Pull request merged**